File: download_files.py
import os
import requests
from datetime import datetime, timedelta
import json
import multiprocessing
import sys

base_url = "https://overrustlelogs.net"
base_path = "./data"

# /Sodapoppin chatlog/January 2019/userlogs/chocolatemilkbandit#17623-17778

os.makedirs("./data/rustle", exist_ok=True)


def download(entry):
    path, uri = entry
    r = requests.get(uri, stream=True)
    if r.status_code == 200:
        with open(path, "wb") as f:
            for chunk in r:
                f.write(chunk)
    return path


def download_cached(entry):
    path, url = entry
    if not os.path.exists(path):
        r = requests.get(url, stream=True)
        if r.status_code == 200:
            with open(path, "wb") as f:
                for chunk in r:
                    f.write(chunk)
    return path


def get_channels():
    download_url = f"{base_url}/api/v1/channels.json"
    path = f"{base_path}/channels.json"
    download((path, download_url))


def get_url_list(channels, days_back, farthest_years):
    urls = []
    for day in range(1, days_back):
        d = datetime.today() - timedelta(days=day)
        date_format = d.strftime("%B %Y/%Y-%m-%d")
        day_stamp = d.strftime("%Y-%m-%d")
        year_stamp = d.strftime("%Y")
        for channel in channels:
            if int(year_stamp) >= int(farthest_years[channel]):
                path = f"{base_path}/rustle/{channel}::{day_stamp}.txt"
                url = f"{base_url}/{channel} chatlog/{date_format}.txt"
                urls.append((path, url))
                print(url)
    return urls


get_channels()

channels = [
    "Destinygg",
    "Trainwreckstv",
    "Forsen",
    "Hasanabi",
    "Pokimane",
    "Lilypichu",
    "Vaushgg",
    "Reckful",
    "Rajjpatel",
    "Whitenervosa",
    "Mrmouton",
    "Greekgodx",
]
farthest_years = {}
for channel in channels:
    r = requests.get(f"{base_url}/api/v1/{channel}/months.json")
    farthest_years[channel] = r.json()[-1][-4:]
print(farthest_years)
url_list = get_url_list(channels, int(sys.argv[1]), farthest_years)
pool = multiprocessing.Pool(processes=20)
pool.map(download_cached, url_list)

File: downloadFiles.js
const fs = require('fs')
const request = require('request')
const util = require('util')
const fg = require('fast-glob')
const {subDays, format} = require('date-fns')
const Promise = require('bluebird')
const R = require('ramda')

const rp = util.promisify(request)
const readFile = util.promisify(fs.readFile)
const writeFile = util.promisify(fs.writeFile)

const baseUrl = 'https://overrustlelogs.net'
const basePath = './data/rustle'

const fullDateFormat = date => format(date, 'MMMM YYYY/YYYY-MM-DD')
const fileDateFormat = date => format(date, 'YYYY-MM-DD')
const yearDateFormat = date => parseInt(format(date, 'YYYY'))

const toPathAndUrl = ({channel, fileDate, fullDate}) => {
  return [
    `${basePath}/${channel}::${fileDate}.txt`,
    `${baseUrl}/${channel} chatlog/${fullDate}.txt`,
  ]
}
const getChannelMonths = async channel => {
  const {body: contents} = await rp({
    uri: `${baseUrl}/api/v1/${channel}/months.json`,
    json: true,
  })
  return [channel, parseInt(contents[contents.length - 1].slice(-4))]
}
const getUrlList = (channels, daysBack, farthestYears) =>
  R.pipe(
    R.map(dayBack => subDays(new Date(), dayBack)),
    R.map(d =>
      [fullDateFormat, fileDateFormat, yearDateFormat].map(fn => fn(d))
    ),
    R.map(([fullDate, fileDate, yearDate]) => {
      const isValidDate = channel => yearDate >= farthestYears[channel]
      const validChannels = channels.filter(channel => isValidDate(channel))
      return validChannels.map(channel =>
        toPathAndUrl({channel, fileDate, fullDate})
      )
    }),
    R.unnest
  )(R.range(1, daysBack))

const downloadFile = async([path, url], json = false) => {
  const res = await rp({uri: url, json})
  if (res.statusCode === 200) {
    await writeFile(path, res.body)
    console.log(`Wrote ${path} to disk.`)
    return path
  } else return false
}

const getFarthestYears = async channels => {
  const farthestYears = await Promise.map(channels, getChannelMonths, {
    concurrency: 20,
  })
  return R.fromPairs(farthestYears)
}

const main = async() => {
  const channelsFile = await readFile('./channels.txt')
  const channels = channelsFile
    .toString()
    .split('\n')
    .filter(x => x !== '')

  const farthestYears = await getFarthestYears(channels)
  console.log(farthestYears)
  const allUrls = getUrlList(channels, process.argv[2] || 10, farthestYears)
  const urlsDownloaded = await fg.async(`${basePath}/*.txt`)
  const urls = allUrls.filter(x => !urlsDownloaded.includes(x[0]))

  console.log(allUrls.length, urlsDownloaded.length, urls.length)
  Promise.map(urls, downloadFile, {concurrency: 20})
}

main()
